core/src/light_protocol/query_handler.rs

```diff
@@ -0,0 +1,458 @@
+            io,
+            peer,
+            msg.as_ref(),
+            SendQueuePriority::High,
+        )?;
+
+        Ok(())
+    }
+
+    fn on_state_root(
+        &self, _io: &NetworkContext, peer: PeerId, rlp: &Rlp,
+    ) -> Result<(), Error> {
+        let resp: GetStateRootResponse = rlp.as_val()?;
+        info!("on_state_root resp={:?}", resp);
+
+        let id = resp.request_id;
+        let (req, sender) = self.match_request::<GetStateRoot>(peer, id)?;
+
+        self.validate_pivot_hash(req.epoch, resp.pivot_hash)?;
+        self.validate_state_root(req.epoch, &resp.state_root)?;
+
+        sender.complete(QueryResult::StateRoot(resp.state_root));
+        // note: in case of early return, `sender` will be cancelled
+
+        Ok(())
+    }
+
+    fn on_get_state_entry(
+        &self, io: &NetworkContext, peer: PeerId, rlp: &Rlp,
+    ) -> Result<(), Error> {
+        let req: GetStateEntry = rlp.as_val()?;
+        info!("on_get_state_entry req={:?}", req);
+        let request_id = req.request_id;
+
+        let pivot_hash = self.get_local_pivot_hash(req.epoch)?;
+        let state_root = self.get_local_state_root(req.epoch)?;
+        let (entry, proof) =
+            self.get_local_state_entry(pivot_hash, &req.key)?;
+        let entry = entry.map(|x| x.to_vec());
+
+        let msg: Box<dyn Message> = Box::new(GetStateEntryResponse {
+            request_id,
+            pivot_hash,
+            state_root,
+            entry,
+            proof,
+        });
+
+        QueryHandler::send_message(
+            io,
+            peer,
+            msg.as_ref(),
+            SendQueuePriority::High,
+        )?;
+
+        Ok(())
+    }
+
+    fn on_state_entry(
+        &self, _io: &NetworkContext, peer: PeerId, rlp: &Rlp,
+    ) -> Result<(), Error> {
+        let resp: GetStateEntryResponse = rlp.as_val()?;
+        info!("on_state_entry resp={:?}", resp);
+
+        let id = resp.request_id;
+        let (req, sender) = self.match_request::<GetStateEntry>(peer, id)?;
+
+        self.validate_pivot_hash(req.epoch, resp.pivot_hash)?;
+        self.validate_state_root(req.epoch, &resp.state_root)?;
+
+        // validate proof
+        if !resp.proof.is_valid(
+            &req.key,
+            &resp.entry,
+            resp.state_root.delta_root,
+            resp.state_root.intermediate_delta_root,
+            resp.state_root.snapshot_root,
+        ) {
+            info!("Invalid proof from peer={}", peer);
+            return Err(ErrorKind::InvalidProof.into());
+        }
+
+        sender.complete(QueryResult::StateEntry(resp.entry));
+        // note: in case of early return, `sender` will be cancelled
+
+        Ok(())
+    }
+
+    fn next_request_id(&self) -> RequestId {
+        let mut id = self.next_request_id.write();
+        let request_id = *id;
+        *id += 1;
+        request_id.into()
+    }
+
+    /// Send `req` to `peer` and wait for result.
+    pub fn execute_request<T: Message + Clone>(
```

Comment:
All messages are handled with limited MIO threads, so we should ensure the method is not time consuming.